x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,739 @@
+package keeper
+
+import (
+	"errors"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	providertypes "github.com/cosmos/interchain-security/x/ccv/provider/types"
+
+	sdkcryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	ccv "github.com/cosmos/interchain-security/x/ccv/types"
+	tmprotocrypto "github.com/tendermint/tendermint/proto/tendermint/crypto"
+)
+
+type VSCID = uint64
+type ProviderPublicKey = tmprotocrypto.PublicKey
+type ConsumerPublicKey = tmprotocrypto.PublicKey
+type ProviderConsAddr = sdk.ConsAddress
+type ConsumerConsAddr = sdk.ConsAddress
+
+func TMCryptoPublicKeyToConsAddr(k tmprotocrypto.PublicKey) sdk.ConsAddress {
+	sdkK, err := sdkcryptocodec.FromTmProtoPublicKey(k)
+	if err != nil {
+		panic("could not get public key from tm proto public key")
+	}
+	return sdk.GetConsAddress(sdkK)
+}
+
+type Store interface {
```

Comment:
Would it be much effort to refactor these abstractions into methods defined for the provider keeper? My understanding of the `/keeper` directory is that it should contain methods defined on the keeper struct, typically state transitions that affect the store. 

I don't necessarily see the value in introducing another layer of abstraction, but maybe I'm missing something

Reply:
The approach taken is to try to encapsulate the business logic of a feature, and the (access to) the state it needs into an isolated unit which can be more easily understood, reasoned about and maintained and tested, without needing additional context

In summary the approach is, in psuedocode

```

// Usage example
k.KeyAssignment(ctx, "foochain").BlaBla()

func (k ProviderKeeper) KeyAssignment(ctx, chainID) KeyAssignment {
    // Return a KeyAssignment instance which encapsulates store access and business logic
}

func (ka KeyAssignment) BlaBla() {
    // business logic
    store.DoStuff()
    // business logic
    store.DoOtherStuff()
}
```

I like this for at least these several reasons

1. It's no longer necessary to pass ctx and chainID everywhere. This reduces the amount of code. Also it reduces the chance of error by substituting chainID for another string type or another string type for a chainID. 
2. By having a KeyAssignment class without full access to the provider keeper it is guaranteed that all of its methods will not call out somewhere else in the codebase through the provider keeper and its friends. This makes it much easier to reason about this feature in isolation.
4. Less setup is needed to test the business logic of the feature. Although I took a meandering path to get the code to where it is, and ended up using a regular KV store in tests, in principle I could have used anything implementing the Store interface which would actually allow very fast in memory read/writes and way more efficient min memory testing. I actually was doing something similar to this very early on when getting the logic right.
5. When writing any code using KeyAssignment the available methods are suggested by the language server, and the suggestions arent cluttered by tons of other methods and fields as would be if using the ProviderKeeper for example.
6. It's possible to find every single place in the code where any key assignment logic is used by simply doing a 'find all references' for the KeyAssignment method on the provider


In my opinion, our codebase generally is implementing the god class antipattern https://en.wikipedia.org/wiki/God_object and I think this is a major contributing factor to our slow development, all the bugs we are finding, and our difficutly testing. I think it would be a good thing if we were using more OOP, encapsulation, single responsiblity principle ect.

I'm always willing to adapt to the needs of the team. If the team feels strongly that things should be done a particular way, I'd like to hear a compelling argument, and I'll be happy to change it.

Reply:
Interesting points and well thought out! I agree with a lot of what you've said here. 

I typically value design consistency to the same extent that I value OOP fundamentals. So it's a trade off for this particular feature. Your decisions all seem reasonable to me.

If we were to redesign cosmos sdk, there could be a generic submodule type or interface in which features like this could implement that interface. One module keeper could manage many submodule types. That way we reap the benefits of everything you've mentioned here, but still have consistency of design between submodules.

Obvi we're committed to the patterns of cosmos sdk for most of this repo, but nice thoughts for the future!

Reply:
@mpoke @jtremback @smarshall-spitzbart Thanks for the feedback guys on this slightly out-of-the-ordinary storage abstraction. I understand that consistency is valued a lot - I will change the code accordingly.

I do recommend that going into the next projects that we try to do things better as I don't think I'm wrong in identifying a major problem with the status quo here.

Thanks